mdata/cwr.go

```diff
@@ -1,23 +1,95 @@
 package mdata
 
 import (
+	"bufio"
+	"bytes"
+	"compress/gzip"
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
 	"time"
 
-	"github.com/grafana/metrictank/mdata/chunk"
 	"github.com/raintank/schema"
 )
 
+type ChunkSaveCallback func()
+
 // ChunkWriteRequest is a request to write a chunk into a store
+//go:generate msgp
 type ChunkWriteRequest struct {
-	Metric    *AggMetric
+	Callback  ChunkSaveCallback `msg:"-"`
 	Key       schema.AMKey
-	Chunk     *chunk.Chunk
 	TTL       uint32
-	Span      uint32
+	T0        uint32
+	Data      []byte
 	Timestamp time.Time
 }
 
 // NewChunkWriteRequest creates a new ChunkWriteRequest
-func NewChunkWriteRequest(metric *AggMetric, key schema.AMKey, chunk *chunk.Chunk, ttl, span uint32, ts time.Time) ChunkWriteRequest {
-	return ChunkWriteRequest{metric, key, chunk, ttl, span, ts}
+func NewChunkWriteRequest(callback ChunkSaveCallback, key schema.AMKey, ttl, t0 uint32, data []byte, ts time.Time) ChunkWriteRequest {
+	return ChunkWriteRequest{callback, key, ttl, t0, data, ts}
+}
+
+// ArchiveRequest is used by the whisper importer utilities to transfer the data
+// from the reader to the writer
+//go:generate msgp
+type ArchiveRequest struct {
+	MetricData         schema.MetricData
+	ChunkWriteRequests []ChunkWriteRequest
+}
+
+func (a *ArchiveRequest) MarshalCompressed() (*bytes.Buffer, error) {
+	var buf bytes.Buffer
+
+	// prefix the buffer with version number 1
+	buf.WriteByte(byte(uint8(1)))
+
+	b, err := a.MarshalMsg(nil)
```

Comment:
No need to marshal to a bytes then pass the byte slice to gzip.  It can be done all at once.

```
err := msgp.Encode(gzip.NewWriter(&buf), a)
``